src/main/java/se/liu/ida/hefquin/engine/queryproc/impl/optimizer/evolutionaryAlgorithm/CostFunctionsForPhysicalPlansImpl.java

```diff
@@ -0,0 +1,117 @@
+package se.liu.ida.hefquin.engine.queryproc.impl.optimizer.evolutionaryAlgorithm;
+
+import se.liu.ida.hefquin.engine.queryplan.PhysicalPlan;
+import se.liu.ida.hefquin.engine.queryproc.QueryOptimizationException;
+import se.liu.ida.hefquin.engine.queryproc.impl.optimizer.utils.CardinalityEstimation;
+
+public class CostFunctionsForPhysicalPlansImpl implements CostFunctionsForPhysicalPlans{
+    protected final CostFunctionsForRootOperatorsImpl costFunctionForRoot;
+
+    public CostFunctionsForPhysicalPlansImpl(CardinalityEstimation cardEstimate) {
+        this.costFunctionForRoot = new CostFunctionsForRootOperatorsImpl(cardEstimate);
+    }
+
+    public CostOfPhysicalPlan getCostOfPhysicalPlan( final PhysicalPlan pp ) throws QueryOptimizationException {
+
+        final int numberOfRequests = getTotalNumberOfRequests(pp);
+        final int shippedRDFTermsForRequests = getTotalShippedRDFTermsForRequests(pp);
+        final int shippedRDFVarsForRequests = getTotalShippedRDFVarsForRequests(pp);
+        final int shippedRDFTermsForResponses = getTotalShippedRDFTermsForResponses(pp);
+        final int shippedRDFVarsForResponses = getTotalShippedVarsForResponses(pp);
+        final int getIntermediateResultsSize = getTotalIntermediateResultsSize(pp);
+
+        final CostOfPhysicalPlan costOfPhysicalPlan = new CostOfPhysicalPlanImpl( numberOfRequests, shippedRDFTermsForRequests , shippedRDFVarsForRequests, shippedRDFTermsForResponses, shippedRDFVarsForResponses, getIntermediateResultsSize);
+
+        return costOfPhysicalPlan;
+    }
+
+    public int getTotalNumberOfRequests( PhysicalPlan pp ) throws QueryOptimizationException {
+        int totalNumberOfRequests = 0;
+        if ( pp.numberOfSubPlans() == 0 ){
+            totalNumberOfRequests += costFunctionForRoot.getNumberOfRequests( pp );
+        }
+
+        for ( int i = 0; i < pp.numberOfSubPlans(); i++ ){
+            final int costOfRootOp = costFunctionForRoot.getNumberOfRequests( pp.getSubPlan(i) );
+            totalNumberOfRequests += costOfRootOp ;
+            getTotalNumberOfRequests( pp.getSubPlan(i) );
```

Comment:
I don't think that this is correct. There are two issues:

1/ The value returned by `getTotalNumberOfRequests( pp.getSubPlan(i) )` must be added to `totalNumberOfRequests`.

2/ After fixing the first issue, lines 35 and 36 become obsolete because the number of requests of the root op of the current subplan is added automatically when calling `getTotalNumberOfRequests( pp.getSubPlan(i) )`.

Reply:
The same issue is also in all the following methods within this class.

Reply:
Hi Olaf, I have fixed these two issues. 
Another TODO is that the cost of **sub**query plans are not cached in the current implementation. 
Can you please check my comments in Class 'CostFunctionsForPhysicalPlansImpl' and let me know what you think?